File: scratch/check_duplicates.py
import geopandas as gpd
import pandas as pd

def check_duplicates():
    geojson_path = 'frontend/public/data/map_data_predictions.geojson'
    gdf = gpd.read_file(geojson_path)
    
    print(f"Total features: {len(gdf)}")
    
    # Check for duplicate village_id
    id_counts = gdf['village_id'].value_counts()
    duplicates_id = id_counts[id_counts > 1]
    
    if not duplicates_id.empty:
        print("\nDuplicate village_ids found:")
        print(duplicates_id)
        # Show some examples
        for vid in duplicates_id.index[:5]:
            print(f"\nExample for village_id {vid}:")
            print(gdf[gdf['village_id'] == vid][['village_name', 'mandal', 'district']])
    else:
        print("\nNo duplicate village_ids found.")
        
    # Check for duplicate village_name + mandal (since names might repeat across mandals)
    gdf['unique_key'] = gdf['village_name'].astype(str) + "_" + gdf['mandal'].astype(str)
    key_counts = gdf['unique_key'].value_counts()
    duplicates_key = key_counts[key_counts > 1]
    
    if not duplicates_key.empty:
        print("\nDuplicate Village+Mandal names found:")
        print(duplicates_key.head(10))
    else:
        print("\nNo duplicate Village+Mandal names found.")

    # Check for IDs that might be mixed up
    # e.g. same name but different IDs
    name_to_ids = gdf.groupby('village_name')['village_id'].unique()
    multiple_ids = name_to_ids[name_to_ids.apply(len) > 1]
    if not multiple_ids.empty:
        print("\nVillages with same name but different IDs:")
        for name, ids in multiple_ids.head(5).items():
            print(f"{name}: {ids}")

if __name__ == "__main__":
    check_duplicates()

File: scratch/check_duplicates.js
const fs = require('fs');
const path = require('path');

const dataDir = path.join(__dirname, '../frontend/public/data');

function checkGeoJson(filename) {
    const p = path.join(dataDir, filename);
    if (!fs.existsSync(p)) return;
    try {
        const data = JSON.parse(fs.readFileSync(p, 'utf8'));
        const features = data.features || [];
        const ids = new Set();
        const names = new Set();
        let duplicatesId = 0;
        let duplicatesName = 0;

        for (const f of features) {
            const props = f.properties || {};
            const id = props.village_id;
            const name = props.village_name || props.Village_Name || props.VILLAGE;
            
            if (id) {
                if (ids.has(id)) duplicatesId++;
                ids.add(id);
            }
            if (name) {
                if (names.has(name)) duplicatesName++;
                names.add(name);
            }
        }
        console.log(`\nFile: ${filename}`);
        console.log(`Total Features: ${features.length}`);
        console.log(`Unique IDs: ${ids.size} (Duplicates: ${duplicatesId})`);
        console.log(`Unique Names: ${names.size} (Duplicates: ${duplicatesName})`);
    } catch(e) {
        console.log(`Error parsing ${filename}: ${e.message}`);
    }
}

function checkJsonList(filename) {
    const p = path.join(dataDir, filename);
    if (!fs.existsSync(p)) return;
    try {
        const data = JSON.parse(fs.readFileSync(p, 'utf8'));
        const list = Array.isArray(data) ? data : (data.features ? data.features : data);
        if (!Array.isArray(list)) {
            console.log(`\nFile: ${filename} - Not a list`);
            return;
        }
        const ids = new Set();
        let duplicatesId = 0;

        for (const row of list) {
            const props = row.properties ? row.properties : row;
            const id = props.village_id;
            if (id) {
                if (ids.has(id)) duplicatesId++;
                ids.add(id);
            }
        }
        console.log(`\nFile: ${filename}`);
        console.log(`Total rows: ${list.length}`);
        console.log(`Unique IDs: ${ids.size} (Duplicates: ${duplicatesId})`);
    } catch(e) {
        console.log(`Error parsing ${filename}: ${e.message}`);
    }
}

['map_data_predictions.geojson', 'map_data_predictions_ntr.geojson', 'village_boundaries.geojson', 'village_boundaries_ntr.geojson', 'villages.geojson'].forEach(checkGeoJson);
['final_dataset.json', 'final_dataset_ntr.json'].forEach(checkJsonList);

